lib/src/command/add.dart

```diff
@@ -0,0 +1,405 @@
+/// the package. The user may pass in a git constraint, host url, or path as
+/// requirements. If no such options are passed in, this command will do a
+/// resolution to find the latest version of the package that is compatible with
+/// the other dependencies in `pubpsec.yaml`, and then enter that as the lower
+/// bound in a ^x.y.z constraint.
+///
+/// Currently supports only adding one dependency at a time.
+class AddCommand extends PubCommand {
+  @override
+  String get name => 'add';
+  @override
+  String get description => 'Add a dependency to pubspec.yaml.';
+  @override
+  String get invocation => 'pub add <package>[:<constraint>] [options]';
+  @override
+  String get docUrl => 'https://dart.dev/tools/pub/cmd/pub-add';
+
+  bool get isDev => argResults['dev'];
+  bool get isDryRun => argResults['dry-run'];
+  String get gitUrl => argResults['git-url'];
+  String get gitPath => argResults['git-path'];
+  String get gitRef => argResults['git-ref'];
+  String get hostUrl => argResults['hosted-url'];
+  String get path => argResults['path'];
+  String get sdk => argResults['sdk'];
+
+  bool get hasGitOptions => gitUrl != null || gitRef != null || gitPath != null;
+  bool get hasHostOptions => hostUrl != null;
+
+  AddCommand() {
+    argParser.addFlag('dev',
+        abbr: 'd',
+        negatable: false,
+        help: 'Adds package to the development dependencies instead.');
+
+    argParser.addOption('git-url', help: 'Git URL of the package');
+    argParser.addOption('git-ref',
+        help: 'Git branch or commit to be retrieved');
+    argParser.addOption('git-path', help: 'Path of git package in repository');
+    argParser.addOption('hosted-url', help: 'URL of package host server');
+    argParser.addOption('path', help: 'Local path');
+    argParser.addOption('sdk', help: 'SDK source for package');
+
+    argParser.addFlag('offline',
+        help: 'Use cached packages instead of accessing the network.');
+
+    argParser.addFlag('dry-run',
+        abbr: 'n',
+        negatable: false,
+        help: "Report what dependencies would change but don't change any.");
+
+    argParser.addFlag('precompile',
+        help: 'Precompile executables in immediate dependencies.');
+  }
+
+  @override
+  Future run() async {
+    if (argResults.rest.isEmpty) {
+      usageException('Must specify a package to be added.');
+    }
+
+    final packageInformation = _parsePackage(argResults.rest.first);
+    final package = packageInformation.first;
+
+    /// Perform version resolution in-memory.
+    final updatedPubSpec =
+        await _addPackageToPubspec(entrypoint.root.pubspec, package);
+
+    SolveResult solveResult;
+
+    try {
+      /// Use [SolveType.UPGRADE] to solve for the highest version of [package]
+      /// in case [package] was already a transitive dependency.
+      solveResult = await resolveVersions(
+        SolveType.UPGRADE,
+        cache,
+        Package.inMemory(updatedPubSpec),
+      );
```

Comment:
Shouldn't we be specifying:
```dart
        lockFile: lockFile,
        useLatest: [package.name],
```

Such that we don't unlock anything, other than the package being added.

So if I have a dependency on `foo` then adding `bar` won't upgrade `foo`, unless upgrading `foo` is required to get some version of `bar`.

Reply:
To my understanding, `useLatest` forces the version of `foo` to be the latest, regardless of other constraints! Consider the following tests (tried with both `SolveType.GET` and `SolveType.UPGRADE`, and the lockfile as mentioned):

1. Dependency override.
```dart
        await servePackages((builder) {
          builder.serve('foo', '1.2.3');
          builder.serve('foo', '1.2.2');
        });

        await d.dir(appPath, [
          d.pubspec({
            'name': 'myapp',
            'dependencies': {},
            'dependency_overrides': {'foo': '1.2.2'}
          })
        ]).create();

        await pubAdd(args: ['foo']);

        await d.cacheDir({'foo': '1.2.2'}).validate();
        await d.appPackagesFile({'foo': '1.2.2'}).validate();
        await d.dir(appPath, [
          d.pubspec({
            'name': 'myapp',
            'dependencies': {'foo': '^1.2.2'},
            'dependency_overrides': {'foo': '1.2.2'}
          })
        ]).validate();
```

which currently passes, but with the addition of `useLatest`, fails with the message `Because myapp depends on foo <1.2.3-∞ or >1.2.3 but the latest version (1.2.3) is required, version solving failed.`, because it forces `foo` to be `1.2.3` even when `1.2.2` exists.

2. 

```dart
  test(
      'empty constraint allows it to choose the latest version not in conflict',
      () async {
    await servePackages((builder) {
      builder.serve('foo', '0.1.0');
      builder.serve('foo', '1.2.3', deps: {'bar': '2.0.4'});
      builder.serve('bar', '2.0.3');
      builder.serve('bar', '2.0.4');
    });

    await d.appDir({'bar': '2.0.3'}).create();

    await pubAdd(args: ['foo']);

    await d.appDir({'foo': '^0.1.0', 'bar': '2.0.3'}).validate();

    await d.cacheDir({'foo': '0.1.0', 'bar': '2.0.3'}).validate();
    await d.appPackagesFile({'foo': '0.1.0', 'bar': '2.0.3'}).validate();
  });
```

which currently passes, but with `useLatest` defined, it forces the usage of `foo:1.2.3`, which conflicts with the pubspec and results in an error:

```bash
[e] SLVR:   fact: foo >=1.2.3 depends on bar 2.0.4
[e] SLVR:   conflict: foo >=1.2.3 depends on bar 2.0.4
[e] SLVR:   ! foo >=1.2.3 is partially satisfied by not foo <1.2.3-∞ or >1.2.3
[e] SLVR:   ! which is caused by "the latest version of foo (1.2.3) is required"
[e] SLVR:   ! thus: every version of foo requires bar 2.0.4
[e] SLVR:   ! not bar 2.0.4 is satisfied by bar 2.0.3
[e] SLVR:   ! which is caused by "myapp depends on bar 2.0.3"
[e] SLVR:   ! thus: foo is forbidden
[e] SLVR:   ! foo any is satisfied by foo any
[e] SLVR:   ! which is caused by "myapp depends on foo any"
[e] SLVR:   ! thus: version solving failed
[e] SLVR: Version solving took 0:00:00.271587 seconds.
[e]     | Tried 1 solutions.
[e] ERR : Because foo >=1.2.3 depends on bar 2.0.4 and the latest version of foo (1.2.3) is required, every version of foo requires bar 2.0.4.
[e]     | So, because myapp depends on both bar 2.0.3 and foo any, version solving failed.
[e] FINE: Exception type: DataException
[e] FINE: package:pub/src/utils.dart 528:35           dataError
[e]     | package:pub/src/command/add.dart 101:7      AddCommand.run
[e]     | ===== asynchronous gap ===========================
[e]     | dart:async                                  Future.catchError
[e]     | package:pub/src/utils.dart 113:52           captureErrors.wrappedCallback
[e]     | package:stack_trace                         Chain.capture
[e]     | package:pub/src/utils.dart 126:11           captureErrors
[e]     | package:pub/src/command_runner.dart 184:13  PubCommandRunner.runCommand
```

As such, unless I'm misunderstanding something about `useLatest`, it doesn't seem like it's suitable parameter for us to set.

Reply:
Interestingly, I just tried running this test:

```dart
  test('upgrades to the latest version not in conflict', () async {
    await servePackages((builder) {
      builder.serve('foo', '1.2.2');
      builder.serve('bar', '2.0.3');
      builder.serve('bar', '2.0.4');
    });

    await d.appDir({'foo': '^1.2.2', 'bar': '2.0.3'}).create();

    await pubGet();

    await d.appDir({'foo': '^1.2.2', 'bar': '2.0.3'}).validate();

    await d.cacheDir({'foo': '1.2.2', 'bar': '2.0.3'}).validate();
    await d.appPackagesFile({'foo': '1.2.2', 'bar': '2.0.3'}).validate();

    globalPackageServer.add((builder) {
      builder.serve('foo', '1.2.3');
      builder.serve('foo', '1.2.4', deps: {'bar': '2.0.4'});
    });

    await pubUpgrade(args: ['foo']);
  });
```

and it resulted in an error (i.e. `pub upgrade foo` did not succeed).

```
Process `pub upgrade` exited with exitCode 1. Output:
[e] FINE: Pub 0.1.2+3
    MSG : Resolving dependencies...
[e] SLVR: fact: myapp is 0.0.0
[e] SLVR: derived: myapp
[e] SLVR: fact: myapp depends on foo ^1.2.2
[e] SLVR: fact: myapp depends on bar 2.0.3
[e] SLVR:   selecting myapp
[e] SLVR:   derived: bar 2.0.3
[e] SLVR:   derived: foo ^1.2.2
[e] IO  : Get versions from http://localhost:34219/api/packages/foo.
[e] IO  : HTTP GET http://localhost:34219/api/packages/foo
[e]     | Accept: application/vnd.pub.v2+json
[e]     | X-Pub-OS: linux
[e]     | X-Pub-Command: upgrade
[e]     | X-Pub-Session-ID: 00BA9360-670E-4EF1-A632-CBF75873A3E3
[e]     | X-Pub-Environment: test-environment
[e]     | X-Pub-Reason: direct
[e]     | user-agent: Dart pub 0.1.2+3
[e] IO  : HTTP response 200 OK for GET http://localhost:34219/api/packages/foo
[e]     | took 0:00:00.053261
[e]     | transfer-encoding: chunked
[e]     | date: Fri, 21 Aug 2020 06:07:14 GMT
[e]     | x-frame-options: SAMEORIGIN
[e]     | content-type: text/plain; charset=utf-8
[e]     | x-xss-protection: 1; mode=block
[e]     | x-content-type-options: nosniff
[e]     | server: dart:io with Shelf
[e] SLVR:   fact: the latest version of foo (1.2.4) is required
[e] SLVR:   derived: not foo <1.2.4-∞ or >1.2.4
[e] IO  : Get versions from http://localhost:34219/api/packages/bar.
[e] IO  : HTTP GET http://localhost:34219/api/packages/bar
[e]     | Accept: application/vnd.pub.v2+json
[e]     | X-Pub-OS: linux
[e]     | X-Pub-Command: upgrade
[e]     | X-Pub-Session-ID: 00BA9360-670E-4EF1-A632-CBF75873A3E3
[e]     | X-Pub-Environment: test-environment
[e]     | X-Pub-Reason: direct
[e]     | user-agent: Dart pub 0.1.2+3
[e] SLVR:     selecting bar 2.0.3
[e] SLVR:     fact: foo >=1.2.4 depends on bar 2.0.4
[e] SLVR:     conflict: foo >=1.2.4 depends on bar 2.0.4
[e] SLVR:     ! foo >=1.2.4 is partially satisfied by not foo <1.2.4-∞ or >1.2.4
[e] SLVR:     ! which is caused by "the latest version of foo (1.2.4) is required"
[e] SLVR:     ! thus: every version of foo requires bar 2.0.4
[e] SLVR:     ! foo any is satisfied by foo ^1.2.2
[e] SLVR:     ! which is caused by "myapp depends on foo ^1.2.2"
[e] SLVR:     ! thus: bar 2.0.4 is required
[e] SLVR:     ! not bar 2.0.4 is satisfied by bar 2.0.3
[e] SLVR:     ! which is caused by "myapp depends on bar 2.0.3"
[e] SLVR:     ! thus: version solving failed
[e] SLVR: Version solving took 0:00:00.147784 seconds.
[e]     | Tried 1 solutions.
[e] FINE: Resolving dependencies finished (0.167s).
[e] ERR : Because foo >=1.2.4 depends on bar 2.0.4 and the latest version of foo (1.2.4) is required, every version of foo requires bar 2.0.4.
[e]     | So, because myapp depends on both foo ^1.2.2 and bar 2.0.3, version solving failed.
[e] FINE: Exception type: SolveFailure
[e] FINE: package:pub/src/solver/version_solver.dart 312:5   VersionSolver._resolveConflict
[e]     | package:pub/src/solver/version_solver.dart 133:27  VersionSolver._propagate
[e]     | package:pub/src/solver/version_solver.dart 97:11   VersionSolver.solve.<fn>
[e]     | ===== asynchronous gap ===========================
[e]     | dart:async                                         Future.catchError
[e]     | package:pub/src/utils.dart 113:52                  captureErrors.wrappedCallback
[e]     | package:stack_trace                                Chain.capture
[e]     | package:pub/src/utils.dart 126:11                  captureErrors
[e]     | package:pub/src/command_runner.dart 184:13         PubCommandRunner.runCommand
```

whereas the following succeeded:

```dart
  test('upgrades to the latest version not in conflict', () async {
    await servePackages((builder) {
      builder.serve('foo', '1.2.2');
      builder.serve('bar', '2.0.3');
      builder.serve('bar', '2.0.4');
    });

    await d.appDir({'foo': '^1.2.2', 'bar': '2.0.3'}).create();

    await pubGet();

    await d.appDir({'foo': '^1.2.2', 'bar': '2.0.3'}).validate();

    await d.cacheDir({'foo': '1.2.2', 'bar': '2.0.3'}).validate();
    await d.appPackagesFile({'foo': '1.2.2', 'bar': '2.0.3'}).validate();

    globalPackageServer.add((builder) {
      builder.serve('foo', '1.2.3');
      builder.serve('foo', '1.2.4', deps: {'bar': '2.0.4'});
    });

    await pubUpgrade();

    await d.appDir({'foo': '^1.2.2', 'bar': '2.0.3'}).validate();

    await d.cacheDir({'foo': '1.2.3', 'bar': '2.0.3'}).validate();
    await d.appPackagesFile({'foo': '1.2.3', 'bar': '2.0.3'}).validate();
  });
```

is this the intended behavior? I would have expected that if the second test passed that the first test would pass too